src/lib.rs

```diff
@@ -519,6 +537,57 @@ impl<W: Watcher, T: Clone + Eq> Watcher for Map<W, T> {
     }
 }
 
+/// Wraps a [`Watcher`] to allow observing a derived value.
+///
+/// See [`Watcher::map`].
+#[derive(derive_more::Debug, Clone)]
+pub struct Filter<W, T>
+where
+    T: Clone + Eq,
+    W: Watcher<Value = T>,
+{
+    #[debug("Arc<dyn Fn(&T) -> bool + 'static>")]
+    filter: Arc<dyn Fn(&T) -> bool + Send + Sync + 'static>,
+    watcher: W,
+    current: Option<T>,
+}
+
+impl<W, T> Watcher for Filter<W, T>
+where
+    T: Clone + Eq,
+    W: Watcher<Value = T>,
+{
+    type Value = Option<T>;
+
+    fn get(&mut self) -> Self::Value {
+        self.current.clone()
+    }
+
+    fn is_connected(&self) -> bool {
+        self.watcher.is_connected()
+    }
+
+    fn poll_updated(
+        &mut self,
+        cx: &mut task::Context<'_>,
+    ) -> Poll<Result<Self::Value, Disconnected>> {
+        loop {
+            let value = ready!(self.watcher.poll_updated(cx)?);
+            let filtered = if (self.filter)(&value) {
+                Some(value)
+            } else {
+                None
+            };
+            if filtered != self.current {
+                self.current = filtered.clone();
+                return Poll::Ready(Ok(filtered));
```

Comment:
This behaves a bit different than I would have expected. Filter returns an Option<TheOriginalValue> instead of just TheOriginalValue, and will trigger even if the value goes from None to Some and Some to None. I would have expected it to return a Watcher<TheOriginalValue> instead.

it is a bit weird for filter to change the type of the stream.

Reply:
The alternative is to only have `filter` on watchables that contain `Option<T>` which is what I had first

Reply:
Most of our watchers have the value wrapped in an option, right?

I am not sure if this is a good API, but I definitely don't think having a fn filter that changes the stream type is intuitive. Maybe it just needs a different name?

Reply:
If you were to use this on our watchers in iroh of type Option<Whatever>, you would get an Option<Option<Whatever>>, and initialized would return immediately, possibly returning a None?

I think having this defined only on Optino<T> is less confusing in comparison.